### PR TITLE
`Development`: Fix charts on course management and course detail page

### LINKED ISSUE #4574: Course Management and course detail pages throw internal server errors
### Describe the bug

If the corresponding course has an amount of active students > 0 in the current week, both pages are throwing internal server errors and do not load the charts.
There seems to be a problem with the aggregation of users for the different weeks on server side (see the server error output below).

### To Reproduce

1. Go to Course administration
2. Inspect whether a chart displays a value > 0 for the current weeks
3. Click on the course card header
4. See the same behavior: Internal Server error is shown and charts do not display data

### Expected behavior

The charts should be loaded appropriately and no server errors should be thrown.

### Screenshots

image
image

### What browsers are you seeing the problem on?

Chrome

### Additional context

_No response_

### Relevant log output

```shell
detail: "Index 4 out of bounds for length 4"
message: "error.http.500"
path: "/api/courses/stats-for-management-overview"
status: 500
title: "Internal Server Error"
type: "https://www.jhipster.tech/problem/problem-with-message"
```

### PR DESCRIPTION
**Pull request by willergeorg**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [ ] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
### Motivation and Context

This PR fixes #4574 
### Description

So far, the aggregation of active users for different weeks did not take into account the following edge case appropriately:
The displayed timespan includes two different years and the most recent week has active users > 0.
The current implementation handles this case somehow by applying an offset to the displayed weeks of the most recent year in order to allocate the correct array position. The problem is, that the offset is hardcoded to 53 calendar weeks which does not necessarily hold for every year (if you are interested, read more about this on e.g. [Wikipedia](https://de.wikipedia.org/wiki/Woche#:~:text=Z%C3%A4hlweise%20nach%20ISO%208601%5BBearbeiten%20%7C%20Quelltext%20bearbeiten%5D)).
This leads currently to internal server errors as the server tries to access an array position that is out of bounds.
I changed the implementation so that it identifies the amount of calendar weeks for the prior year and applies the appropriate offset in case of such a situation as described above.
### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course with activity in the most recent week

1. Log in to Artemis
2. Navigate to Course Administration
3. Make sure no internal server error is thrown and the charts are filled with data
4. Click on the course card and check the same for the chart on the course detail page.

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Screenshots

Screenshots documenting the current state are provided by the connected issue. The diagrams should now work again:
image
image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/CourseService.java [hunk 1]
```diff
@@ -443,7 +443,12 @@ else if (!usersInSameSlot.contains(username)) {
             int dateWeek = getWeekOfDate(date);
             int startDateWeek = getWeekOfDate(startDate);
             int weeksDifference;
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/CourseService.java:445]:
```suggestion
```

File: src/main/java/de/tum/in/www1/artemis/service/CourseService.java [hunk 2]
```diff
@@ -443,7 +443,12 @@ else if (!usersInSameSlot.contains(username)) {
...
-            weeksDifference = dateWeek < startDateWeek ? dateWeek == startDateWeek - 1 ? 0 : dateWeek + 53 - startDateWeek : dateWeek - startDateWeek;
+            // we need to take into account that years can have 53 or 52 calendar weeks
+            ZonedDateTime firstDateOfStartYear = ZonedDateTime.of(startDate.getYear(), 1, 1, 0, 0, 0, 0, startDate.getZone());
+            ZonedDateTime lastDateOfStartYear = ZonedDateTime.of(startDate.getYear(), 12, 31, 0, 0, 0, 0, startDate.getZone());
+            boolean beginsOrEndsOnThursday = firstDateOfStartYear.getDayOfWeek() == DayOfWeek.THURSDAY || lastDateOfStartYear.getDayOfWeek() == DayOfWeek.THURSDAY;
+            int lastCalendarWeek = beginsOrEndsOnThursday ? 53 : 52;
+            weeksDifference = dateWeek < startDateWeek ? dateWeek == startDateWeek - 1 ? 0 : dateWeek + lastCalendarWeek - startDateWeek : dateWeek - startDateWeek;
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/CourseService.java:451]:
```suggestion
            int weeksDifference = dateWeek < startDateWeek ? dateWeek == startDateWeek - 1 ? 0 : dateWeek + lastCalendarWeek - startDateWeek : dateWeek - startDateWeek;
```
With the double ternary operator it is also quite hard to understand what is going on in this line. Can you change it into a nested if-else, please?

**Review by b-fein — CHANGES_REQUESTED** [review]
Is there a difference between the new calculation and using the provided standard library functionality instead?
```java
int weeksDifference = (int) ChronoUnit.WEEKS.between(startDate, date);
```

In any case, `StatisticsRepository::mergeResultsIntoArrayForQuarter` needs the exact same fix. Can you apply it there as well (ideally without duplicating code if the standard library cannot be used)?

**Comment by akesfeden** [comment]
First of all, thanks for spotting and working on this bug, I also have failing tests because of it.

I agree with @b-fein on using `ChronoUnit.WEEKS.between(startDate, date)` (or equivalent `startDate.until(date, ChronoUnit.WEEKS)`) if it is applicable.

**Comment by willergeorg** [comment]
First of all, thanks @b-fein and @akesfeden for taking the time and reviewing my changes. Unfortunately, there were some devs that worked on this problem in parallel without knowing someone else already has a look on it. As @krusche already fixes this problem in #4571 and additionally reduces the code duplication there, this PR is obsolete.

**Pull request closed without merging**